### PR TITLE
wire up all prometheus proxy config settings for flow and the converters

### LINKED ISSUE #5110: Update HTTPClientConfig for new fields available in Prometheus configuration
Our current Prometheus dependency has introduced some extra proxy-related fields grouped under the [ProxyConfig](https://github.com/prometheus/common/blob/94bf9828e56d9670579b28a9f78237d3cd8d0395/config/http_config.go#L1178) struct.

We should include the new fields in the Validate and Convert methods of our HTTPClientConfig, and update the documentation of all components making use of HTTPClientConfig.

### LINKED ISSUE #6261: Custom HTTP headers for discovery.http
### Request

I would like to set custom HTTP headers in `discovery.http`

### Use case

I'm getting Prometheus targets from Netbox using this plugin: https://github.com/FlxPeters/netbox-plugin-prometheus-sd
Netbox requires the `Authorization: Token xxxx` header for authentication.  My Netbox system is behind an Apache proxy which uses `Authorization` for another purpose, so I rename the header using an Apache config like this:
```
RequestHeader set Authorization "expr=%{req:Netbox-Authorization}"
```

Therefore I need to set the `Netbox-Authorization` header.

### PR DESCRIPTION
**Pull request by erikbaranowski**
#### PR Description

Wires in the 3 additional proxy configurations from promtheus to the http client config and elsewhere. I improved some code along the way and called out discrepancies between docs and code.

#### Which issue(s) this PR fixes

Closes #6261
Closes #5110

#### PR Checklist

- [x] CHANGELOG.md updated
- [x] Documentation added
- [x] Tests updated
- [x] Config converters updated

### CODE REVIEW

File: component/common/config/types.go
Comment (erikbaranowski) [inline component/common/config/types.go:35]: This refactor pushes the responsibility for validating other types to their own Validate funcs except for ProxyConfig (squashed). It also tries to more sanely handle the scenarios where only 1 auth method can be set. I think this code got more and more convoluted over time.

File: component/common/config/types.go
Comment (erikbaranowski) [inline component/common/config/types.go:159]: These mirror Prometheus' ProxyConfig validations.

https://github.com/prometheus/common/blob/773d5664eb8d228d2c6ba235b1af12fbf7ee9020/config/http_config.go#L1210

File: component/common/config/types_test.go
Comment (erikbaranowski) [inline component/common/config/types_test.go:231]: In some cases, the error changed because the order of validation will be different. I don't think this is harmful.

File: component/discovery/azure/azure.go
Comment (erikbaranowski) [inline component/discovery/azure/azure.go:77]: Need to explicitly validate squashed properties

File: component/discovery/docker/docker_test.go
Comment (erikbaranowski) [inline component/discovery/docker/docker_test.go:30]: These error messages got longer to encapsulate all potential overlapping auths.

File: component/discovery/linode/linode_test.go
Comment (erikbaranowski) [inline component/discovery/linode/linode_test.go:36]: Prove out that an undocumented block could be set on the component.

File: docs/sources/flow/reference/components/discovery.azure.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/discovery.azure.md:1]: This component brings in portions of http client config and not the whole thing.

File: docs/sources/flow/reference/components/discovery.dockerswarm.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/discovery.dockerswarm.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the bearer* args and update the component args to match.

File: docs/sources/flow/reference/components/discovery.hetzner.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/discovery.hetzner.md:48]: I tried to sync this wording everywhere. It is inconsistent on many components.

File: docs/sources/flow/reference/components/discovery.kubelet.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/discovery.kubelet.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the bearer* args and update the component args to match.

File: docs/sources/flow/reference/components/discovery.lightsail.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/discovery.lightsail.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the bearer* args and update the component args to match.

File: docs/sources/flow/reference/components/discovery.linode.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/discovery.linode.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the bearer* args and update the component args to match.

File: docs/sources/flow/reference/components/discovery.marathon.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/discovery.marathon.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the bearer* args and update the component args to match.

File: docs/sources/flow/reference/components/loki.source.podlogs.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/loki.source.podlogs.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the newly documented args and update the component args to match.

File: docs/sources/flow/reference/components/prometheus.operator.podmonitors.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/prometheus.operator.podmonitors.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the newly documented args and update the component args to match.

File: docs/sources/flow/reference/components/prometheus.operator.servicemonitors.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/prometheus.operator.servicemonitors.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the newly documented args and update the component args to match.

File: docs/sources/flow/reference/components/remote.kubernetes.configmap.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/remote.kubernetes.configmap.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the newly documented args and update the component args to match.

File: docs/sources/flow/reference/components/remote.kubernetes.secret.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/remote.kubernetes.secret.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the newly documented args and update the component args to match.

File: component/discovery/consul/consul_test.go
Comment (erikbaranowski) [inline component/discovery/consul/consul_test.go:38]: Prove that the cascading validation is happening for properties on HttpClientConfig

File: component/common/config/types.go
Comment (erikbaranowski) [inline component/common/config/types.go:61]: A Validate setting fields makes me uncomfortable, but I already touched a ton of code on this PR.

File: component/common/config/types.go
Comment (erikbaranowski) [inline component/common/config/types.go:266]: A Validate setting fields makes me uncomfortable, but I already touched a ton of code on this PR.

File: converter/internal/promtailconvert/testdata/docker_relabel.river
Comment (rfratto) [inline converter/internal/promtailconvert/testdata/docker_relabel.river:22]: This is a little unfortunate, but not a blocker; can you spend like 15 minutes on this to see if we can stop it from generating this?

File: docs/sources/shared/flow/reference/components/http-client-proxy-config-description.md
Comment (rfratto) [inline docs/sources/shared/flow/reference/components/http-client-proxy-config-description.md:14]: It feels a little funny for us to describe something called `no_proxy` and then say "if you configure `no_proxy`, you must configure `proxy_url`." Maybe flipping the order of these would be useful? 

IMO, the way to describe `no_proxy` is as a denylist of IPs/CIDR notation/domain names where the proxy server _will not_ be used, and a connection is instead made directly (which is why you have to specify the `proxy_url` if you're also configuring where not to proxy to).

File: docs/sources/shared/flow/reference/components/http-client-proxy-config-description.md
Comment (rfratto) [inline docs/sources/shared/flow/reference/components/http-client-proxy-config-description.md:16]: This is throwing me off a bit, especially with the mixed casing. We might want to copy Go's phrasing here:

> ProxyFromEnvironment returns the URL of the proxy to use for a given request, as indicated by the environment variables HTTP_PROXY, HTTPS_PROXY and NO_PROXY (or the lowercase versions thereof). Requests use the proxy from the environment variable matching their scheme, unless excluded by NO_PROXY.

**Review by rfratto — APPROVED** [review]
LGTM % some final comments and pending a review from @clayton-cornell :)

File: docs/sources/flow/reference/components/discovery.azure.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/discovery.azure.md:36]:
```suggestion
`no_proxy`               | `string`            | Comma-separated list of IP addresses, CIDR notations, and domain names to exclude from proxying.                         |                      | no
```

We could be a little more explicit what the list is here. If this is OK, we could cascade this to the rest of the tables.

File: docs/sources/flow/reference/components/discovery.azure.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/discovery.azure.md:37]:
```suggestion
`proxy_from_environment` | `bool`              | Use the proxy URL indicated by environment variables.                      | `false`              | no
```

If adding `the` here makes sense, this can also be cascaded through the rest of the tables

File: docs/sources/flow/reference/components/discovery.hetzner.md
Comment (clayton-cornell) [inline docs/sources/flow/reference/components/discovery.hetzner.md:48]:
```suggestion
 At most, one of the following can be provided:
```

I only tagged this one instance with the suggestion for adding a comma. It's probably easier to search/replace in the branch vs trying to find every instance of this and post a suggested change to add a comma.

Comment (clayton-cornell) [reply docs/sources/flow/reference/components/discovery.hetzner.md:48]: We need to insert a comma after `most`.

File: docs/sources/shared/flow/reference/components/http-client-proxy-config-description.md
Comment (clayton-cornell) [inline docs/sources/shared/flow/reference/components/http-client-proxy-config-description.md:13]:
```suggestion
`no_proxy` can contain IPs, CIDR notations, and domain names. IP and domain names can contain port numbers.
```

**Review by clayton-cornell — COMMENTED** [review]
Some comments. Since they apply to all doc files, I commented once with a suggestion. It's a lot easier to search/replace in your local branch vs a large number of identical suggested changes.

File: docs/sources/flow/reference/components/prometheus.operator.probes.md
Comment (erikbaranowski) [inline docs/sources/flow/reference/components/prometheus.operator.probes.md:1]: This component uses the whole http client config but had undocumented args. If the docs were correct then I can remove the newly documented args and update the component args to match.

**Review by mattdurham — APPROVED** [review]
Kubelet change seems reasonable though I doubt it actually supports all the auth parameters , but I guess that’s part of sharing. Either way this is better

File: docs/sources/flow/reference/components/discovery.marathon.md
Comment (wildum) [inline docs/sources/flow/reference/components/discovery.marathon.md:35]: I think that the bearer token is not supported by marathon from what I understand [here](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#marathon_sd_config): 
# Optional `Authorization` header configuration.
# NOTE: The current version of DC/OS marathon (v1.11.0) does not support
# standard `Authentication` header, use `auth_token` or `auth_token_file`
# instead.

Comment (erikbaranowski) [reply docs/sources/flow/reference/components/discovery.marathon.md:35]: We reached consensus offline to keep the whole HTTPClientConfig wired in on the component and document it. If future versions of marathon begin to support these authentication methods we will not need to update the component.

**Review by clayton-cornell — APPROVED** [review]

**Review by tpaschalis — APPROVED** [review]
Huge chunk of work, nice one @erikbaranowski 🙌

**Pull request merged**
